### PR TITLE
fork-tree: add support for find_node_mut_where

### PR DESCRIPTION
**Pull request by sorpaas**
(Split out of #4600, for easier review/merge.)

This adds support to get mutable references of `Node` in a `ForkTree`, by using a helper method `find_node_index_where` that returns the indexes to access the node. Mutable reference is needed for consensus such as Sassafras, because we need to push VRF proofs after the epoch data is created.

### CODE REVIEW

File: utils/fork-tree/src/lib.rs [hunk 1]
```diff
@@ -250,6 +250,31 @@ impl<H, N, V> ForkTree<H, N, V> where
 		Ok(None)
 	}
 
+	/// Same as `find_node_where`, but returns mutable reference.
```
Comment (bkchr) [inline utils/fork-tree/src/lib.rs]:
```suggestion
	/// Same as [`find_node_where`](Self::find_node_where), but returns mutable reference.
```

File: utils/fork-tree/src/lib.rs [hunk 2]
```diff
@@ -655,14 +684,78 @@ mod node_implementation {
...
+				},
+			}
+		}
+
+		/// Find a node in the tree that is the deepest ancestor of the given
+		/// block hash which also passes the given predicate, backtracking
+		/// when the predicate fails.
+		/// The given function `is_descendent_of` should return `true` if the second hash (target)
+		/// is a descendent of the first hash (base).
+		pub fn find_node_mut_where<F, P, E>(
```
Comment (bkchr) [inline utils/fork-tree/src/lib.rs]:
```suggestion
		pub fn find_node_where_mut<F, P, E>(
```

File: utils/fork-tree/src/lib.rs [hunk 3]
```diff
@@ -250,6 +250,31 @@ impl<H, N, V> ForkTree<H, N, V> where
 		Ok(None)
 	}
 
+	/// Same as `find_node_where`, but returns mutable reference.
+	pub fn find_node_mut_where<F, E, P>(
```
Comment (bkchr) [inline utils/fork-tree/src/lib.rs]:
```suggestion
	pub fn find_node_where_mut<F, E, P>(
```

File: utils/fork-tree/src/lib.rs [hunk 4]
```diff
@@ -609,16 +634,17 @@ mod node_implementation {
 		/// when the predicate fails.
 		/// The given function `is_descendent_of` should return `true` if the second hash (target)
 		/// is a descendent of the first hash (base).
-		// FIXME: it would be useful if this returned a mutable reference but
-		// rustc can't deal with lifetimes properly. an option would be to try
-		// an iterative definition instead.
-		pub fn find_node_where<F, P, E>(
+		///
+		/// The returned indexes are from last to first, meaning the last is the least significant
```
Comment (rphmeier) [inline utils/fork-tree/src/lib.rs]:
```suggestion
		/// The returned indices are from last to first, meaning the last is the least significant
```

Comment (rphmeier) [reply utils/fork-tree/src/lib.rs]: it's not clear what "least significant" means in this comment, but from reading the code i take it to be the earlier index in the traversal path.

Comment (sorpaas) [reply utils/fork-tree/src/lib.rs]: I updated the docs just calling it the "earliest" and "final" index.

**Comment by rphmeier** [comment]
Thank you Wei!

**Pull request merged**